grails-app/controllers/au/org/ala/ecodata/CommentController.groovy

```diff
@@ -0,0 +1,161 @@
+package au.org.ala.ecodata
+
+import grails.converters.JSON
+import groovy.json.JsonSlurper
+
+class CommentController {
+    def outputService
+
+    def ignores = ["action","controller"]
+
+    def list() {
+        List comments = []
+        def sort = params.sort ? params.sort : "dateCreated"
+        def order = params.order ? params.order :  "desc"
+        def offset = params.start ? params.start : 0
+        def max = params.pageSize ? params.pageSize : 10
+        Boolean sortOrder = order == 'asc'? true : false;
+        String  entityId = params.entityId
+        String entityType = params.entityType
+        Integer total = 0;
+
+        if( !entityId || !entityType){
+            response.sendError(400, 'Insufficient parameters provided. Missing either entityId or entityType')
+        } else {
+//           todo: is there a better way to do count. is the below method potentially slow?
+            total = Comment.findAllWhere([ 'entityId':entityId, 'entityType':entityType, parent: null]).size();
+            Comment.findAllWhere(['entityId':entityId, 'entityType':entityType, parent: null], [sort:sort,order:order,offset:offset,max:max]).each {
+                Map mapOfProperties = outputService.getCommentProperties(it)
+                comments.add(mapOfProperties)
+            }
+
+            comments.each { comment ->
+                if(comment.children?.size()){
+                    outputService.sortCommentChildren(comment.children, sortOrder);
+                }
+            }
+
+            render text:  [total: total, items: comments] as JSON, contentType: 'application/json'
+        }
+    }
+
+
+    @RequireApiKey
+    def create(){
+        def jsonSlurper = new JsonSlurper()
+        def json = jsonSlurper.parse(request.getReader())
+        if (!json.userId){
+            response.sendError(400, 'Missing userId')
+        } else if (!json.entityId || !json.entityType){
+            response.sendError(400, 'Missing entityId and/or entityType')
+        } else if ( !json.text ){
+            response.sendError(400, 'Missing text');
+        } else {
+            Comment l = new Comment(json)
+            Comment parent;
+            if(l.dateCreated == null){
+                l.dateCreated = new Date();
+            }
+
+            Comment comment = l.save(true)
```

Comment:
Data updates should generally be done in the service tier, where you can run in a transaction. It doesn't make much difference here since we're not using transactions, bit it's a good practise to follow. Controllers are for processing request/response data. Services are for performing business rules.